server/src/main/java/org/elasticsearch/index/mapper/DocumentLeafReader.java

```diff
@@ -0,0 +1,476 @@
+                visitor.binaryField(fieldInfo, data);
+            }
+        }
+    }
+
+    @Override
+    public CacheHelper getCoreCacheHelper() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Terms terms(String field) throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public NumericDocValues getNormValues(String field) throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Bits getLiveDocs() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public PointValues getPointValues(String field) throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void checkIntegrity() throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public LeafMetaData getMetaData() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Fields getTermVectors(int docID) throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int numDocs() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int maxDoc() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CacheHelper getReaderCacheHelper() {
+        throw new UnsupportedOperationException();
+    }
+
+    // Our StoredFieldsVisitor implementations only check the name of the passed-in
+    // FieldInfo, so that's the only value we need to set here.
+    private static FieldInfo fieldInfo(String name) {
+        return new FieldInfo(
+            name,
+            0,
+            false,
+            false,
+            false,
+            IndexOptions.NONE,
+            DocValuesType.NONE,
+            -1,
+            Collections.emptyMap(),
+            0,
+            0,
+            0,
+            false
+        );
+    }
+
+    private static NumericDocValues numericDocValues(List<Number> values) {
+        if (values.size() == 0) {
+            return null;
+        }
+        return new NumericDocValues() {
+            @Override
+            public long longValue() {
+                return values.get(0).longValue();
+            }
+
+            @Override
+            public boolean advanceExact(int target) {
+                return true;
+            }
+
+            @Override
+            public int docID() {
+                return 0;
+            }
+
+            @Override
+            public int nextDoc() {
+                return 0;
+            }
+
+            @Override
+            public int advance(int target) {
+                return 0;
+            }
```

Comment:
We might expose ourselves to subtle bugs by not implementing the DocIdSetIterator contract correctly. Let's make sure that `docID()` returns -1 when the iterator is not positioned and make `nextDoc`/`advance` return `NO_MORE_DOCS` when appropriate?